deps/jemalloc/src/jemalloc.c

```diff
@@ -3927,5 +3927,145 @@ jemalloc_postfork_child(void) {
+        phn_merge(extent_t, ph_link, r0, r1, extent_snad_comp, bin->slabs_nonfull.ph_root);
+        /* Clear the temp heap after merge */
+        extent_heap_new(&bin->slabs_nonfull_temp);
+        assert(extent_heap_empty(&bin->slabs_nonfull_temp));
+        bin->initing_defrag = false;
+    }
 }
+
+JEMALLOC_EXPORT void JEMALLOC_NOTHROW
+init_defrag(void) {
+    unsigned long long i;
+    tsd_t *tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+        const bin_info_t *bin_info = &bin_infos[i];
+        uint64_t used_slabs = bin->stats.nonfull_slabs + (bin->slabcur ? 1 : 0);
+        uint64_t full_slabs = bin->stats.curslabs - bin->stats.nonfull_slabs - (bin->slabcur ? 1 : 0);
+        uint64_t needed_slabs = (bin->stats.curregs - full_slabs * bin_info->nregs) / bin_info->nregs;
+        if ((bin->stats.curregs - full_slabs * bin_info->nregs) % bin_info->nregs) needed_slabs += 1;
+        needed_slabs += full_slabs;
+        used_slabs += full_slabs;
+        assert(bin->highest_slab_to_retain_inited == false);
+        assert(extent_heap_empty(&bin->slabs_nonfull_temp));
+        if (used_slabs != needed_slabs) {
+            int64_t nonfull_to_retain = needed_slabs - full_slabs;
+            if (bin->slabcur) nonfull_to_retain--;
+            bin->defrag_slabs_to_retain = nonfull_to_retain;
+            bin->initing_defrag = true;
+        } else {
+            bin->defrag_slabs_to_retain = 0;
+            bin->initing_defrag = false;
+        }
+        malloc_mutex_unlock(tsd_tsdn(tsd), &bin->lock);
+    }
+}
+
+JEMALLOC_EXPORT int JEMALLOC_NOTHROW
+init_defrag_step(long long max_time) {
+    unsigned long long i;
+    tsd_t * tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+    long long start_time = ustime();
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+        if (bin->initing_defrag) {
+            init_defrag_bin_step(bin, start_time, max_time);
+            /* If we're still initing after the step it means we've timed out but have more work */
+            if (bin->initing_defrag) {
+                malloc_mutex_unlock(tsd_tsdn(tsd), &bin->lock);
+                break;
+            }
+        }
+        malloc_mutex_unlock(tsd_tsdn(tsd), &bin->lock);
+    }
+    return i < SC_NBINS;
+}
+
+JEMALLOC_EXPORT void JEMALLOC_NOTHROW
+finish_defrag(void) {
+    unsigned long long i;
+    tsd_t *tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+
+        if (bin->initing_defrag) {
+            extent_t *r0 = extent_heap_first(&bin->slabs_nonfull);
```

Comment:
i think its better to encapsulate the merge call into one function, maybe in the ph.h. (basically the three 5 lines here)

Reply:
fixed